### ISSUE TITLE
Should Integers, Naturals etc subclass Range?

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#17856]
In the same way that `Reals` is a subclass of `Interval` I think it might be useful for `Integers`, `Naturals`, `Naturals0` to be subclasses of `Range`.

The advantage of doing this is that it unifies the logic for handling intersections etc. With `multipledispatch` we potentially need code for the full Cartesian product of different `Set` class combinations. Almost all of the code currently in the `Naturals`, `Naturals0` and `Integers` classes would be unneeded if we did this and most other references to them in intersect/union code could go as well.

We could make it so that e.g. `Range(oo)` automatically evaluates to `Naturals0` so that when relevant the results of intersections, unions etc. comes back as one of the standard sets.

The only code I can see that doesn't quite work with Range is `Integers.__iter__` which currently produces `0,1,-1,2,-1,...` whereas `Range(-oo,oo).__iter__` currently raises. I don't know if it makes sense to define iteration for a `Range` with an infinite start and I'm actually unsure whether there is any particular use in iterating over `Integers` anyway (apart from the fun of it :) )

**Comment by smichr** [comment]
> apart from the fun of it

If you want to see a few specific instances of an ImageSet defined over Integers you would need this iteration, wouldn't you?

**Comment by oscarbenjamin** [comment]
You wouldn't *need* `__iter__` for that but yes I see that it could be useful now.

The question is then whether it's okay to have that one `Range` iterate differently from the way that others iterate. Iteration/ordering is a peculiarity of `Range` as a `Set` that can't generally be preserved under unions etc. but is there because `Range` has alternate uses like in codegen. The alternate uses wouldn't be able to do anything with `Range(-oo, oo)` though so maybe it doesn't matter if this particular `Range` iterates differently.

**Comment by asmeurer** [comment]
I think iterating over the integers is important actually. It lets you do things like iterate the solutions of `solveset(sin(x), x, Reals)`. But I don't think it would be a big deal if we enabled iteration over ranges with infinite start. I think a relatively bigger consequence of this is that Naturals and so on would get some of the Range methods, like slicing.

**Comment by oscarbenjamin** [comment]
We could disallow slicing of Naturals, Naturals0 in their subclass code. I think it's fine for them to be sliceable though.

**Comment by sylee957** [comment]
One confusion that `Range` can introduce when used with other set objects is that it has an additional information about the direction how it enumerates.

For example `Range(0, 5)` and `Range(4, -1, -1)` should be considered different in the context of `Range`, and how would the equality, union, intersection should be defined between such objects.

So, I would introduce a new object like `RangeSet`, which can indicate a subset of integers from A to B, but with no information about the direction.

**Comment by oscarbenjamin** [comment]
> So, I would introduce a new object like `RangeSet`, which can 'genuinely' be a set of some integers bound by a start, a stop, and a step, but is purely a set.

I agree that a new set class should be used instead of Range. However I wonder if it needs to be a 3-argument Range type object or if it can just be `IntegerInterval` and always have a step-size of 1. The `Range` class is very complicated because of the difficulty in handling the step parameter but in sets I don't think it is needed. With `IntegerInterval` the other step-sizes can be created using `ImageSet`.

**Comment by asmeurer** [comment]
From my memory of working on Range a while back (and maybe those who have worked on it more recently can comment too), it might actually be cleaner if Range were split into two classes, one for finite ranges and one for infinite ranges. A lot of logic in the Range code has to deal with the differences between these two cases, and a lot of the complexity of ranges with symbolic values comes from handling the complexity of a value potentially being infinite. If there were separate FiniteRange and InfiniteRange classes, it could end up simplifying the code a lot. Although perhaps to really simplify things we would have to break backwards compatibility so that `Range(n)` with a symbolic `n` can only mean `FiniteRange`.

I also wonder if Range itself can be simplified by internally representing it as an ImageSet or ConditionSet on something simpler, like Range(n). That's assuming that the logic for ImageSet is smart enough to handle all the operations like contains, intersection, and so on. 

Range has three things that complicate it:

- The step, which makes the size of the range dependent on `(end - start) % step`. The step can also be negative which inverts almost all logic. 
- The fact that Range is 0-based and half-open. Particularly the half-open, meaning that `end` is never included in the Range, makes things very difficult to reason about. 
- The fact that the start or end can be infinite, which necessarily changes the logic. For instance, if the start is infinite, then last element is `end - step` (usually the last element depends on `(end - start) % step`). And visa-versa if the step is negative. 

One must consider in particular symbolic ranges, but the logic in the code is more or less the logic for the symbolic case, because they have to deal with `start`, `end`, and `step` as variables. So making things work when they are symbolic just means using `Piecewise` instead of `if` and so on.

For example, try to write down the formula for the last element of a Range (`Range(start, end, step)[-1]`). I can't do it off the top of my head for finite ranges, because of the step (and negative step) logic, and if you add the infinite range logic in there, it is even worse. 

If we could eliminate the infinite range branching, so that we always know the range is either finite or infinite, and then further fallback to ImageSet/ConditionSet (not sure which) logic, that could make things much better. Just look at the code for [`Range.__getitem__`](https://github.com/sympy/sympy/blob/d19dfaeaa93de6d4fc0462d7eb1d47f681be1d15/sympy/sets/fancysets.py#L732-L863) or [Range.intersect(Range)](https://github.com/sympy/sympy/blob/d19dfaeaa93de6d4fc0462d7eb1d47f681be1d15/sympy/sets/handlers/intersection.py#L105-L215). There's got to be a better way to do this.

I don't actually think the ordering is a huge deal. Integers and Naturals have already have orderings too, because `__iter__` is defined for them. The ordering only means that if we use ImageSet we have to define the image function carefully so that the correct ordering is used.

**Comment by oscarbenjamin** [comment]
I think that only having a step size of +1 and eliminating open vs closed boundaries can simplify things a huge amount. Having infinite start and end isn't a significant problem once you get rid of those other two complexities.

For example given `IntegerRange(a, b).intersect(IntegerRange(c, d))` the result is just `IntegerRange(Max(a, c), Min(b, d))` a simple formula that works for finite numbers, symbolic ranges and infinite start/stop as well.

The `__getitem__` method wouldn't be needed for a set-like `IntegerInterval`. The ordering of range is a problem because particularly when combined with the mixed closed/open ends it complicates calculating basic things like the maximal element of the set.

**Comment by asmeurer** [comment]
> Having infinite start and end isn't a significant problem once you get rid of those other two complexities.

Maybe, but I think having different cases unavoidable once you add a step, because the definition for infinite start with step > 1 has to be different than for finite ranges. Unless there's some clever way to write the formulas that works for both.

**Comment by asmeurer** [comment]
Anyway, if the idea behind IntegerRange is to be closed on both ends (unless infinite), I think that's useful, although it's basically equivalent to `Integers.intersect(Interval(a, b))`.

**Comment by sylee957** [comment]
I find one other problem with the current set framework that there is no way to represent integers multiplied with `n`.
But I think that this kind of stuff can be either done via extending `Range(-oo, oo, n)` or adding a new class like `MultiplieInteger`

I want to hear the idea about whether to pursue adding

1. `IntegerInterval` without step and `MultipleInteger`.
2. `IntegerInterval` with step.

should be better.
